File: src/kf.py
import numpy as np


class KalmanFilter(object):

    def __init__(self, state_dim, measurement_dim):
        if state_dim < 1:
            raise ValueError('state_dim must be 1 or greater')
        if measurement_dim < 1:
            raise ValueError('measurement_dim must be 1 or greater')

        self._state_dim = state_dim                             # Number of state variables
        self._measurement_dim = measurement_dim                 # Number of measurement variables
        self._x = np.zeros((state_dim, 1))                      # State mean
        self._P = np.eye(state_dim)                             # State covariance
        self._Q = np.eye(state_dim)                             # Process covariance
        self._F = np.eye(state_dim)                             # State transition matrix
        self._H = np.zeros((measurement_dim, state_dim))        # Measurement function
        self._K = np.zeros((state_dim, measurement_dim))        # Kalman gain
        self._y = np.zeros((measurement_dim, 1))                # Residual
        self._S = np.zeros((measurement_dim, measurement_dim))  # System uncertainty

    @property
    def x(self):
        return self._x

    @x.setter
    def x(self, x):
        if x.shape != (self._state_dim, 1):
            raise ValueError('x shape is wrong')
        self._x = x

    @property
    def P(self):
        return self._P

    @P.setter
    def P(self, P):
        if P.shape != (self._state_dim, self._state_dim):
            raise ValueError('P shape is wrong')
        self._P = P

    @property
    def Q(self):
        return self._Q

    @Q.setter
    def Q(self, Q):
        if Q.shape != (self._state_dim, self._state_dim):
            raise ValueError('Q shape is wrong')
        self._Q = Q

    @property
    def F(self):
        return self._F

    @F.setter
    def F(self, F):
        if F.shape != (self._state_dim, self._state_dim):
            raise ValueError('F shape is wrong')
        self._F = F

    @property
    def H(self):
        return self._H

    @ H.setter
    def H(self, H):
        if H.shape != (self._measurement_dim, self._state_dim):
            raise ValueError('H shape is wrong')
        self._H = H

    @property
    def K(self):
        return self._K

    @ K.setter
    def K(self, K):
        if K.shape != (self._state_dim, self._measurement_dim):
            raise ValueError('K shape is wrong')
        self._K = K

    @property
    def y(self):
        return self._y

    @ y.setter
    def y(self, y):
        if y.shape != (self._measurement_dim, 1):
            raise ValueError('y shape is wrong')
        self._y = y

    @property
    def S(self):
        return self._S

    @ S.setter
    def S(self, S):
        if S.shape != (self._measurement_dim, self._measurement_dim):
            raise ValueError('S shape is wrong')
        self._S = S

    def predict(self):
        self._x = np.dot(self._F, self._x)
        self._P = np.dot(self._F, self._P).dot(self._F.T) + self._Q
        return self._x, self._P

    def update(self, z: np.ndarray, R: np.ndarray):
        if z.shape != (self._measurement_dim, 1):
            raise ValueError('z shape is wrong')
        if R.shape != (self._measurement_dim, self._measurement_dim):
            raise ValueError('R shape is wrong')

        PHT = np.dot(self._P, self._H.T)
        self._S = np.dot(self._H, PHT) + R
        self._K = np.dot(PHT, np.linalg.inv(self._S))
        self._y = z - np.dot(self._H, self._x)
        self._x = self._x + np.dot(self._K, self._y)
        self._P = self._P - np.dot(self._K, self._H).dot(self._P)  # Should be P = (I - KH)P, but might be unstable?
        return self._x, self._P

File: src/kf.cpp
#include "kf.hpp"

#include <cassert>

namespace kf {

KalmanFilter::KalmanFilter(int state_dim, int measurement_dim) :
  state_dim_{state_dim},
  measurement_dim_{measurement_dim}
{
  assert(state_dim > 0);
  assert(measurement_dim > 0);

  x_ = Eigen::MatrixXd::Zero(state_dim, 1);
  P_ = Eigen::MatrixXd::Identity(state_dim, state_dim);
  Q_ = Eigen::MatrixXd::Identity(state_dim, state_dim);
  F_ = Eigen::MatrixXd::Identity(state_dim, state_dim);
  H_ = Eigen::MatrixXd::Zero(measurement_dim, state_dim);
  K_ = Eigen::MatrixXd::Zero(state_dim, measurement_dim);
  y_ = Eigen::MatrixXd::Zero(measurement_dim, 1);
  S_ = Eigen::MatrixXd::Zero(measurement_dim, measurement_dim);
}

void KalmanFilter::set_x(const Eigen::MatrixXd &x)
{
  assert(x.rows() == state_dim_ && x.cols() == 1);
  x_ = x;
}

void KalmanFilter::set_P(const Eigen::MatrixXd &P)
{
  assert(P.rows() == state_dim_ && P.cols() == state_dim_);
  P_ = P;
}

void KalmanFilter::set_Q(const Eigen::MatrixXd &Q)
{
  assert(Q.rows() == state_dim_ && Q.cols() == state_dim_);
  Q_ = Q;
}

void KalmanFilter::set_F(const Eigen::MatrixXd &F)
{
  assert(F.rows() == state_dim_ && F.cols() == state_dim_);
  F_ = F;
}

void KalmanFilter::set_H(const Eigen::MatrixXd &H)
{
  assert(H.rows() == measurement_dim_ && H.cols() == state_dim_);
  H_ = H;
}

void KalmanFilter::set_K(const Eigen::MatrixXd &K)
{
  assert(K.rows() == state_dim_ && K.cols() == measurement_dim_);
  K_ = K;
}

void KalmanFilter::set_y(const Eigen::MatrixXd &y)
{
  assert(y.rows() == measurement_dim_ && y.cols() == 1);
  y_ = y;
}

void KalmanFilter::set_S(const Eigen::MatrixXd &S)
{
  assert(S.rows() == measurement_dim_ && S.cols() == measurement_dim_);
  S_ = S;
}

void KalmanFilter::predict()
{
  x_ = F_ * x_;
  P_ = F_ * P_ * F_.transpose() + Q_;

#if 1
  assert(x_.rows() == state_dim_ && x_.cols() == 1);
  assert(P_.rows() == state_dim_ && P_.cols() == state_dim_);
#endif
}

void KalmanFilter::update(const Eigen::MatrixXd &z, const Eigen::MatrixXd &R)
{
  assert(z.rows() == measurement_dim_ && z.cols() == 1);
  assert(R.rows() == measurement_dim_ && R.cols() == measurement_dim_);

  Eigen::MatrixXd PHT = P_ * H_.transpose();
  S_ = H_ * PHT + R;
  K_ = PHT * S_.inverse();
  y_ = z - H_ * x_;
  x_ = x_ + K_ * y_;
  P_ = P_ - K_ * H_ * P_;  // Should be P = (I - KH)P, but might be unstable?

#if 1
  assert(x_.rows() == state_dim_ && x_.cols() == 1);
  assert(P_.rows() == state_dim_ && P_.cols() == state_dim_);
#endif
}

} // namespace kf